crates/mun_codegen/src/ir/body.rs

```diff
@@ -259,16 +260,53 @@ impl<'a, 'b, D: IrDatabase> BodyIrGenerator<'a, 'b, D> {
     }
 
     /// Generates an IR value that represents the given `Literal`.
-    fn gen_literal(&mut self, lit: &Literal) -> BasicValueEnum {
+    fn gen_literal(&mut self, lit: &Literal, expr: ExprId) -> BasicValueEnum {
         match lit {
-            Literal::Int(v) => self
-                .db
-                .context()
-                .i64_type()
-                .const_int(unsafe { mem::transmute::<i64, u64>(*v) }, true)
-                .into(),
-
-            Literal::Float(v) => self.db.context().f64_type().const_float(*v as f64).into(),
+            Literal::Int(v) => {
+                let ty = match &self.infer[expr] {
+                    hir::Ty::Apply(hir::ApplicationTy {
+                        ctor: hir::TypeCtor::Int(int_ty),
+                        ..
+                    }) => int_ty,
+                    _ => unreachable!(
+                        "cannot construct an IR value for anything but an integral type"
+                    ),
+                };
+
+                let context = self.db.context();
+                let has_sign = ty.signedness == hir::Signedness::Signed;
+                let ir_ty = match ty.resolve(&self.db.target_data_layout()).bitness {
+                    hir::IntBitness::X8 => context.i8_type().const_int(v.value as u64, has_sign),
+                    hir::IntBitness::X16 => context.i16_type().const_int(v.value as u64, has_sign),
+                    hir::IntBitness::X32 => context.i32_type().const_int(v.value as u64, has_sign),
+                    hir::IntBitness::X64 => context.i64_type().const_int(v.value as u64, has_sign),
+                    hir::IntBitness::X128 => {
+                        context.i128_type().const_int(v.value as u64, has_sign)
```

Comment:
This is a narrowing cast. If this is the case, then the `max` function should not allow more than `std::u64::max` for `u128`

Reply:
Good find, fixed and added test.